src/Misc/khronosTextureContainer2.ts

```diff
@@ -1,69 +1,227 @@
 
-                internalTexture.width = internalTexture.baseWidth = ktxTexture.baseWidth;
-                internalTexture.height = internalTexture.baseHeight = ktxTexture.baseHeight;
-                internalTexture.generateMipMaps = false;
+        if (KhronosTextureContainer2._WorkerPoolPromise) {
+            return KhronosTextureContainer2._WorkerPoolPromise.then((workerPool) => {
+                return new Promise((resolve, reject) => {
+                    workerPool.push((worker, onComplete) => {
+                        const onError = (error: ErrorEvent) => {
+                            worker.removeEventListener("error", onError);
+                            worker.removeEventListener("message", onMessage);
+                            reject(error);
+                            onComplete();
+                        };
 
-                const result = ktxTexture.glUpload();
-                if (result.error === 0) {
-                    internalTexture._webGLTexture = result.texture;
-                }
-                else {
-                    throw new Error(`Failed to upload: ${result.error}`);
-                }
+                        const onMessage = (message: MessageEvent) => {
+                            if (message.data.action === "decoded") {
+                                worker.removeEventListener("error", onError);
+                                worker.removeEventListener("message", onMessage);
+                                if (!message.data.success) {
+                                    reject({ message: message.data.msg });
+                                } else {
+                                    try {
+                                        this._createTexture(message.data.decodedData, internalTexture);
+                                        resolve();
+                                    } catch (err) {
+                                        reject({ message: err });
+                                    }
+                                }
+                                onComplete();
+                            }
+                        };
+
+                        worker.addEventListener("error", onError);
+                        worker.addEventListener("message", onMessage);
+
+                        // note: we can't transfer the ownership of data.buffer because if using a fallback texture the data.buffer buffer will be used by the current thread
+                        worker.postMessage({ action: "decode", data, caps: compressedTexturesCaps }/*, [data.buffer]*/);
+                    });
+                });
+            });
+        }
 
-                internalTexture.isReady = true;
+        return new Promise((resolve, reject) => {
+            if (!KhronosTextureContainer2._Ktx2Decoder) {
+                KhronosTextureContainer2._Ktx2Decoder = new KTX2DECODER.KTX2Decoder();
             }
-            finally {
-                ktxTexture.delete();
+
+            try {

```

Comment:
try/catch in a promise executor isn't necessary. This happens automatically.

Reply:
I just performed the test: if I raise an exception in `KTX2Decoder.decode`, it is caught by the `catch` of the `try / catch` construct, not by the `.catch` construct of the Promise.

Or maybe you meant something else?

Reply:
It should. This works for me? https://jsfiddle.net/Lhg3dfc7/

Reply:
You throw the error inside the promise, whereas KTX2Decoder.decode is not inside a promise (it returns a promise).

Reply:
I typically handle this by creating a dummy Promise.resolve().then(): https://jsfiddle.net/7s3bcmfo/

Reply:
Ok, performed the changes.